app/TuistApp/Sources/Views/DevicesView/DevicesViewModel.swift

```diff
@@ -122,24 +143,46 @@ final class DevicesViewModel: Sendable {
         try? appStorage.set(PinnedSimulatorsKey.self, value: pinnedSimulators)
     }
 
+    func refreshDevices() async throws {
+        isRefreshing = true
+
+        do {
+            try await onAppear()
+        } catch {
+            isRefreshing = false
+            throw error
+        }
+
+        isRefreshing = false
+    }
+
     func onAppear() async throws {
         devices = try await deviceController.findAvailableDevices()
 
         let simulators = try await simulatorController.devicesAndRuntimes()
             .sorted()
 
-        if let selectedDevice = try appStorage.get(SelectedDeviceKey.self) {
-            self.selectedDevice = selectedDevice
-        } else {
-            selectedDevice = simulators.first(where: { !$0.device.isShutdown }).map { .simulator($0) }
-        }
-
         pinnedSimulators = try appStorage.get(PinnedSimulatorsKey.self)
 
         unpinnedSimulators = Set(simulators)
             .subtracting(Set(pinnedSimulators))
             .map { $0 }
             .sorted()
+
+        func getSelectedDevice(reduced device: ReducedDevice) -> SelectedDevice? {
+            switch device {
+            case let .simulator(id):
+                simulators.first(where: { $0.id == id }).map { .simulator($0) }
+            case let .device(id):
+                devices.first(where: { $0.id == id }).map { .device($0) }
+            }
+        }
```

Comment:
I would avoid defining a function inside another function – I would define it at the class level.

I'd also rewrite getting the selected device from storage with the following method:
```swift
func storedSelectedDevice() -> SelectedDevice? {
  guard let selectedDevice = try? appStorage.get(SelectedDeviceKey.self) else { return nil }
  switch selectedDevice {
            case let .simulator(id):
                simulators.first(where: { $0.id == id }).map { .simulator($0) }
            case let .device(id):
                devices.first(where: { $0.id == id }).map { .device($0) }
  } 
}
```

Reply:
This is a good improvement, but we’ll need to define an additional `simulators` variable. Should we change `unpinnedSimulators` to be a computed property instead?